lib/matplotlib/legend.py

```diff
@@ -409,42 +409,27 @@ def _set_loc(self, loc):
         # _legend_box will draw itself at the location of the return
         # value of the find_offset.
         self._loc_real = loc
-        if loc == 0:
-            _findoffset = self._findoffset_best
-        else:
-            _findoffset = self._findoffset_loc
-
-#       def findoffset(width, height, xdescent, ydescent):
-#           return _findoffset(width, height, xdescent, ydescent, renderer)
-
-        self._legend_box.set_offset(_findoffset)
-
-        self._loc_real = loc
         self.stale = True
 
     def _get_loc(self):
         return self._loc_real
 
     _loc = property(_get_loc, _set_loc)
 
-    def _findoffset_best(self, width, height, xdescent, ydescent, renderer):
-        "Helper function to locate the legend at its best position"
-        ox, oy = self._find_best_position(width, height, renderer)
-        return ox + xdescent, oy + ydescent
-
-    def _findoffset_loc(self, width, height, xdescent, ydescent, renderer):
-        "Helper function to locate the legend using the location code"
+    def _findoffset(self, width, height, xdescent, ydescent, renderer):
+        "Helper function to locate the legend"
 
-        if iterable(self._loc) and len(self._loc) == 2:
-            # when loc is a tuple of axes(or figure) coordinates.
-            fx, fy = self._loc
-            bbox = self.get_bbox_to_anchor()
-            x, y = bbox.x0 + bbox.width * fx, bbox.y0 + bbox.height * fy
-        else:
+        if self._loc == 0:  # "best".
+            x, y = self._find_best_position(width, height, renderer)
+        elif self._loc in Legend.codes.values():  # Fixed location.
             bbox = Bbox.from_bounds(0, 0, width, height)
             x, y = self._get_anchored_bbox(self._loc, bbox,
                                            self.get_bbox_to_anchor(),
                                            renderer)
+        else:  # Axes or figure coordinates.
```

Comment:
The previous version would raise AssertionError both for invalid single values (loc=99) and for invalid length values (loc=(1, 2, 3)).  The new implementation now raises TypeError for the former and ValueError for the latter.  I think the new version is better, given that most functions can indeed raise both TypeError and ValueError when given an incorrect input, so you need to catch both anyways (whereas catching AssertionError is pretty rare).